engine/client/cl_spray.c

```diff
@@ -0,0 +1,442 @@
+			continue;
+
+		if( w * h < SPRAY_MAX_SURFACE )
+		{
+			*width = w;
+			*height = h;
+			return;
+		}
+	}
+	// fallback: minimal size
+	*width = 16;
+	*height = 16;
+}
+
+// generates mipmaps for indexed data
+static void CL_GenerateMipmaps( const byte *source, int width, int height, byte *mip1, byte *mip2, byte *mip3 )
+{
+	const int sizes[3][2] = {
+		{ width / 2, height / 2 },
+		{ width / 4, height / 4 },
+		{ width / 8, height / 8 }
+	};
+	byte      *mips[3] = { mip1, mip2, mip3 };
+	int m, mw, mh, step, y, x;
+
+	for( m = 0; m < 3; ++m )
+	{
+		if( !mips[m] )
+			continue;
+		mw = sizes[m][0];
+		mh = sizes[m][1];
+		step = 1 << ( m + 1 );
+		for( y = 0; y < mh; ++y )
+		{
+			for( x = 0; x < mw; ++x )
+			{
+				mips[m][y * mw + x] = source[( y * step ) * width + ( x * step )];
+			}
+		}
+	}
+}
+
+// loads and prepares the image
+static rgbdata_t *CL_LoadAndPrepareImage( const char *inputFile, int *width, int *height, rgbdata_t **scaled_image_out )
+{
+	rgbdata_t *image = FS_LoadImage( inputFile, NULL, 0 );
+	rgbdata_t *scaled_image = NULL;
+	qboolean  resampled = false;
+	byte      *resampled_buf = NULL;
+	byte      *pix = NULL;
+	int i, bpp = 0; // bytes per pixel
+
+	if( !image )
+		return NULL;
+
+	bpp = ( image->type == PF_RGBA_32 ) ? 4 : 3;
+
+	if( bpp )
+	{
+		for( i = 0; i < image->width * image->height; i++ )
+		{
+			pix = &image->buffer[i * bpp];
+			if( bpp == 4 )
+			{
+				if( pix[3] <= 128 ) // transparent
+				{
+					pix[0] = 0;
+					pix[1] = 0;
+					pix[2] = 255;
+					pix[3] = 255; // or 0, doesn't matter
+				}
+			}
+			else // bpp == 3
+			{
+				if( pix[0] == 0 && pix[1] == 0 && pix[2] == 255 )
+					pix[2] = 254;
+			}
+		}
+	}
+
+	*width = image->width;
+	*height = image->height;
+	CL_AdjustSprayDimensions( width, height );
+	if( *width < 16 || *height < 16 || *width % 16 != 0 || *height % 16 != 0 )
+	{
+		Con_DPrintf( S_ERROR "Invalid dimensions after adjustment: %dx%d\n", *width, *height );
+		FS_FreeImage( image );
+		return NULL;
+	}
+	if( *width != image->width || *height != image->height )
+	{
+		resampled_buf = Image_ResampleInternal( image->buffer, image->width, image->height, *width, *height, image->type, &resampled );
+		if( !resampled_buf )
+		{
+			Con_DPrintf( S_ERROR "Failed to create scaled image\n" );
+			FS_FreeImage( image );
+			return NULL;
+		}
+		scaled_image = Mem_Malloc( host.imagepool, sizeof( rgbdata_t ));
+		scaled_image->buffer = Mem_Malloc( host.imagepool, *width * *height * 4 );
+		scaled_image->width = *width;
+		scaled_image->height = *height;
+		scaled_image->type = image->type;
+		scaled_image->size = *width * *height * 4;
+		memcpy( scaled_image->buffer, resampled_buf, *width * *height * 4 );
+		Con_DPrintf( S_GREEN "Created scaled image %dx%d\n", *width, *height );
+	}
+	*scaled_image_out = scaled_image;
+	return image;
+}
+
+// write wad3 file
+static qboolean CL_WriteSprayWAD3( const byte *indexed, int width, int height, const byte *palette, int lump_type )
```

Comment:
Ideally, such things must be implemented in `imagelib`.

Maybe add handler for `.wad` extension in `imagelib` that will load and save tempdecal?